src/legacy/ui/public/agg_types/controls/number_interval.tsx

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { get } from 'lodash';
+import React, { useEffect } from 'react';
+
+import { EuiFieldNumber, EuiFormRow, EuiIconTip } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { FormattedMessage } from '@kbn/i18n/react';
+import { AggParamEditorProps } from '../../vis/editors/default';
+
+const label = (
+  <>
+    <FormattedMessage
+      id="common.ui.aggTypes.numberInterval.minimumIntervalLabel"
+      defaultMessage="Minimum interval"
+    />{' '}
+    <EuiIconTip
+      position="right"
+      content={
+        <FormattedMessage
+          id="common.ui.aggTypes.numberInterval.minimumIntervalTooltip"
+          defaultMessage="Interval will be automatically scaled in the event that the provided value creates more buckets than specified by Advanced Setting's {histogramMaxBars}"
+          values={{ histogramMaxBars: 'histogram:maxBars' }}
+        />
+      }
+      type="questionInCircle"
+    />
+  </>
+);
+
+function NumberIntervalParamEditor({
+  agg,
+  editorConfig,
+  isInvalid,
+  value,
+  setTouched,
+  setValidity,
+  setValue,
+}: AggParamEditorProps<number | undefined>) {
+  const base: number = get(editorConfig, 'interval.base');
+  const min = base || 0;
+
+  useEffect(
```

Comment:
It seems like we should have a single source of truth for the "setValidity" logic:

```js
// Compute this, instead of destructuring it
const isValid = (value !== undefined && value >= min);

// onChange simply sets value
const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
  const newValue = parseFloat(e.target.value);
  setValue(isNaN(newValue) ? undefined : newValue);
};

// This effect has the sole responsibility of setting the validity
useEffect(() => { setValidity(isValid) }, [isValid]);
```

If we do this, `setValidity` logic lives in one place; the effect. It's a value that we compute on each render, and we only really need to refer to the computed value within this component, so I think we can just leave out the `isInvalid` destructured argument, and use `!isValid` instead.

Reply:
Thank you for the comment. One source of truth for `setValidity` really makes sense.

> ...we can just leave out the isInvalid destructured argument, and use !isValid instead.

We have a case when a user opens the form at the first time (e.g. selects `Histogram` aggregation). The field value and the number interval value are empty, but they are not in red until the user touches an input or presses `Apply` button. That is user-friendly behavior for form validation.
To handle that case we make the validation visible after user action. Since we cannot detect applying the form in the `NumberIntervalParamEditor`, we handle this above, in the [wrapper](https://github.com/elastic/kibana/blob/master/src/legacy/ui/public/vis/editors/default/agg_param.js#L106),  using `isInvalid` prop.

_...in 5 minutes_

@chrisdavies your comment helped me to see inconsistency in the way how validation is displayed. So I did some refactoring and now there is `showValidation` prop which indicates if validation should be hidden or not. And `isValid` value is calculated within the component. I also updated other migrated controls. I like how it looks now. :)